### ISSUE TITLE
[FEAT] Option for automatic type-casting in `datediff_level`

### ISSUE THREAD

**Issue by ADBond** [moj-analytical-services/splink#1030]
### Is your proposal related to a problem?
Sometimes I may be working with date columns which are formatted as strings, which may be useful if e.g. I want to have a mix of datediff and levenshtein levels (although for duckdb at least levenshtein works on dates), or I don't have ready control of my input columns/want to make work reproducible between parties with only settings (i.e. not requiring source data modification)

### Describe the solution you'd like
Flag to `datediff_level` comparison level, something like `cll.datediff_level("dob", 1, "year", cast_strings_to_date=True)`, and similar for the comparison `datediff_at_thresholds`

### Describe alternatives you've considered
Can be worked around by using an explicit date column, or hand-writing the relevant SQL.

### Additional context
Would need to consider the case that there is a chance that the data quality is not verified and may contain invalid dates (e.g. `'1981-04-31`) - probably would want this to act as automatically `FALSE` for date comparisons, but be able to levenshtein on it (could well be a typo for `'1981-04-30'`). Might require some use of `TRY_CAST` and checking for `NULL` values in the result.

**Comment by samnlindsay** [comment]
I foresee various problems arising by not enforcing one type, or being too lenient with allowing strings.

Sidenote: What happens if we create a linker for two datasets with DOB columns of different types?

Potential edge cases to consider with a string column:
- `"2023-01-01"` vs `"not a date"` - should be captured by `null_level`, not `else_level`
- `"not a date"` vs `"not a date"` - should be captured by `null_level`, not `exact_match_level`
- `"2023-01-01"` vs `"1/1/23"` - should be captured by `exact_match_level`, not `else_level`

If we don't want to enforce date type on columns provided to datediff comparisons, then I think we need to ensure both sides are formatted similarly, or formatted at all. Casting strings to date only allows for dates formatted as `YYYY-MM-DD` right? 

Basically, you can easily turn dates to strings if you want to use string operations, but it's a pain to do the reverse.

**Comment by RossKen** [comment]
Closed by #1162
